src/Database/LSMTree/Internal/BlobRef.hs

```diff
@@ -186,3 +188,15 @@ readBlobIOOp buf bufoff
       (fromIntegral blobSpanOffset :: FS.FileOffset)
       buf (FS.BufferOffset bufoff)
       (fromIntegral blobSpanSize :: FS.ByteCount)
+
+-- | A stronger alternative to 'cast', which can be used to cast /only/ the @h@
+-- parameter to 'WeakBlobRef'.
+class CastWeakBlobRef m where
+  castWeakBlobRef ::
+       (Typeable h, Typeable h')
+    => WeakBlobRef m h
+    -> Maybe (WeakBlobRef m h')
+
+instance CastWeakBlobRef IO where
+  castWeakBlobRef = cast
+  {-# INLINE castWeakBlobRef #-}
```

Comment:
`castWeakBlobRef` can be implemented for any `m`, so we don't need a class:

```haskell
castWeakBlobRef r = fmap (\Refl -> r) (eqT :: Maybe (h :~: h'))
```

Reply:
I was not doing the type fiddling quite as nicely as @mheinzel is doing it. I've changed `checkBlobRefType` to use `eqT` instead